In [1]:
import pandas as pd

In [166]:
# loading a csv file 

coffee = pd.read_csv("https://raw.githubusercontent.com/KeithGalli/complete-pandas-tutorial/refs/heads/master/warmup-data/coffee.csv")

In [167]:
# to view the file
coffee

Day Coffee Type  Units Sold
0      Monday    Espresso          25
1      Monday       Latte          15
2     Tuesday    Espresso          30
3     Tuesday       Latte          20
4   Wednesday    Espresso          35
5   Wednesday       Latte          25
6    Thursday    Espresso          40
7    Thursday       Latte          30
8      Friday    Espresso          45
9      Friday       Latte          35
10   Saturday    Espresso          45
11   Saturday       Latte          35
12     Sunday    Espresso          45
13     Sunday       Latte          35

In [168]:
# provides statistical description of the dataframe
coffee.describe()

Units Sold
count   14.000000
mean    32.857143
std      9.346798
min     15.000000
25%     26.250000
50%     35.000000
75%     38.750000
max     45.000000

In [169]:
#  Provides a quick summary of the DataFrame—shows column names, data types, non-null counts, and memory usage.

coffee.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Day          14 non-null     object
 1   Coffee Type  14 non-null     object
 2   Units Sold   14 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 468.0+ bytes


In [170]:
# to return first n datas of the dataframe
coffee.head(5)

Day Coffee Type  Units Sold
0     Monday    Espresso          25
1     Monday       Latte          15
2    Tuesday    Espresso          30
3    Tuesday       Latte          20
4  Wednesday    Espresso          35

In [171]:
# to return last n datas of the dataframe
coffee.tail(3)

Day Coffee Type  Units Sold
11  Saturday       Latte          35
12    Sunday    Espresso          45
13    Sunday       Latte          35

In [172]:
# to get random datas
coffee.sample(8,  random_state = 1) #renadom_state is used so that the values doesn't change

Day Coffee Type  Units Sold
3     Tuesday       Latte          20
7    Thursday       Latte          30
6    Thursday    Espresso          40
2     Tuesday    Espresso          30
10   Saturday    Espresso          45
4   Wednesday    Espresso          35
1      Monday       Latte          15
12     Sunday    Espresso          45

In [173]:
coffee

Day Coffee Type  Units Sold
0      Monday    Espresso          25
1      Monday       Latte          15
2     Tuesday    Espresso          30
3     Tuesday       Latte          20
4   Wednesday    Espresso          35
5   Wednesday       Latte          25
6    Thursday    Espresso          40
7    Thursday       Latte          30
8      Friday    Espresso          45
9      Friday       Latte          35
10   Saturday    Espresso          45
11   Saturday       Latte          35
12     Sunday    Espresso          45
13     Sunday       Latte          35

In [174]:
# to access specific value
# loc - label based data selection

coffee.loc[0:4,["Day","Units Sold"]] #slicing

Day  Units Sold
0     Monday          25
1     Monday          15
2    Tuesday          30
3    Tuesday          20
4  Wednesday          35

In [175]:
# iloc = index based data selection

coffee.iloc[0:4,[0,2]]

# .iloc[index:index,[column,column]]

Day  Units Sold
0   Monday          25
1   Monday          15
2  Tuesday          30
3  Tuesday          20

In [176]:
# coffee.index = coffee["Day"]
# coffee.loc["Monday":"Wednesday"]

In [177]:
# modify
coffee.loc[1, "Units Sold"] = 30
coffee

Day Coffee Type  Units Sold
0      Monday    Espresso          25
1      Monday       Latte          30
2     Tuesday    Espresso          30
3     Tuesday       Latte          20
4   Wednesday    Espresso          35
5   Wednesday       Latte          25
6    Thursday    Espresso          40
7    Thursday       Latte          30
8      Friday    Espresso          45
9      Friday       Latte          35
10   Saturday    Espresso          45
11   Saturday       Latte          35
12     Sunday    Espresso          45
13     Sunday       Latte          35

In [178]:
# to get specific value

coffee.at[1,"Units Sold"] #get single value  : label based

np.int64(30)

In [179]:
coffee.iat[5,1] #index

'Latte'

In [180]:
# get a specific column
coffee["Day"]

0        Monday
1        Monday
2       Tuesday
3       Tuesday
4     Wednesday
5     Wednesday
6      Thursday
7      Thursday
8        Friday
9        Friday
10     Saturday
11     Saturday
12       Sunday
13       Sunday
Name: Day, dtype: object

In [181]:
# to sort data

coffee.sort_values(["Units Sold","Coffee Type"],ascending=[0,1]) 

# coffee.sort_values("Units Sold")
# coffee.sort_values("Units Sold",ascending = False) descending order

Day Coffee Type  Units Sold
8      Friday    Espresso          45
10   Saturday    Espresso          45
12     Sunday    Espresso          45
6    Thursday    Espresso          40
4   Wednesday    Espresso          35
9      Friday       Latte          35
11   Saturday       Latte          35
13     Sunday       Latte          35
2     Tuesday    Espresso          30
1      Monday       Latte          30
7    Thursday       Latte          30
0      Monday    Espresso          25
5   Wednesday       Latte          25
3     Tuesday       Latte          20

In [182]:
# filtering

coffee.loc[coffee["Units Sold"]>27][["Coffee Type"]]


Coffee Type
1        Latte
2     Espresso
4     Espresso
6     Espresso
7        Latte
8     Espresso
9        Latte
10    Espresso
11       Latte
12    Espresso
13       Latte

In [183]:
coffee[(coffee["Units Sold"] > 28) & (coffee["Coffee Type"] == "Espresso")] #returns how much units of espresso has been sold 

Day Coffee Type  Units Sold
2     Tuesday    Espresso          30
4   Wednesday    Espresso          35
6    Thursday    Espresso          40
8      Friday    Espresso          45
10   Saturday    Espresso          45
12     Sunday    Espresso          45

In [184]:
# add extra column
coffee_new = coffee.copy()
coffee_new["Price"] = 4.99

In [185]:
coffee

Day Coffee Type  Units Sold
0      Monday    Espresso          25
1      Monday       Latte          30
2     Tuesday    Espresso          30
3     Tuesday       Latte          20
4   Wednesday    Espresso          35
5   Wednesday       Latte          25
6    Thursday    Espresso          40
7    Thursday       Latte          30
8      Friday    Espresso          45
9      Friday       Latte          35
10   Saturday    Espresso          45
11   Saturday       Latte          35
12     Sunday    Espresso          45
13     Sunday       Latte          35

In [186]:
coffee_new

Day Coffee Type  Units Sold  Price
0      Monday    Espresso          25   4.99
1      Monday       Latte          30   4.99
2     Tuesday    Espresso          30   4.99
3     Tuesday       Latte          20   4.99
4   Wednesday    Espresso          35   4.99
5   Wednesday       Latte          25   4.99
6    Thursday    Espresso          40   4.99
7    Thursday       Latte          30   4.99
8      Friday    Espresso          45   4.99
9      Friday       Latte          35   4.99
10   Saturday    Espresso          45   4.99
11   Saturday       Latte          35   4.99
12     Sunday    Espresso          45   4.99
13     Sunday       Latte          35   4.99

In [187]:
import numpy as np

coffee["newprice"]=np.where(coffee["Coffee Type"]=="Espresso",3.99,5.99)
coffee

Day Coffee Type  Units Sold  newprice
0      Monday    Espresso          25      3.99
1      Monday       Latte          30      5.99
2     Tuesday    Espresso          30      3.99
3     Tuesday       Latte          20      5.99
4   Wednesday    Espresso          35      3.99
5   Wednesday       Latte          25      5.99
6    Thursday    Espresso          40      3.99
7    Thursday       Latte          30      5.99
8      Friday    Espresso          45      3.99
9      Friday       Latte          35      5.99
10   Saturday    Espresso          45      3.99
11   Saturday       Latte          35      5.99
12     Sunday    Espresso          45      3.99
13     Sunday       Latte          35      5.99

In [188]:
# tp drop column

coffee.drop(columns = ["Price"], inplace = True) #inplace : returns modified copies


KeyError: "['Price'] not found in axis"

In [189]:
coffee

Day Coffee Type  Units Sold  newprice
0      Monday    Espresso          25      3.99
1      Monday       Latte          30      5.99
2     Tuesday    Espresso          30      3.99
3     Tuesday       Latte          20      5.99
4   Wednesday    Espresso          35      3.99
5   Wednesday       Latte          25      5.99
6    Thursday    Espresso          40      3.99
7    Thursday       Latte          30      5.99
8      Friday    Espresso          45      3.99
9      Friday       Latte          35      5.99
10   Saturday    Espresso          45      3.99
11   Saturday       Latte          35      5.99
12     Sunday    Espresso          45      3.99
13     Sunday       Latte          35      5.99

In [190]:
# revenue
coffee["revenue"] = coffee["Units Sold"] * coffee["newprice"]
coffee

Day Coffee Type  Units Sold  newprice  revenue
0      Monday    Espresso          25      3.99    99.75
1      Monday       Latte          30      5.99   179.70
2     Tuesday    Espresso          30      3.99   119.70
3     Tuesday       Latte          20      5.99   119.80
4   Wednesday    Espresso          35      3.99   139.65
5   Wednesday       Latte          25      5.99   149.75
6    Thursday    Espresso          40      3.99   159.60
7    Thursday       Latte          30      5.99   179.70
8      Friday    Espresso          45      3.99   179.55
9      Friday       Latte          35      5.99   209.65
10   Saturday    Espresso          45      3.99   179.55
11   Saturday       Latte          35      5.99   209.65
12     Sunday    Espresso          45      3.99   179.55
13     Sunday       Latte          35      5.99   209.65

In [191]:
# rename

coffee.rename(columns = {"newprice":"price"}, inplace = True)

In [192]:
coffee

Day Coffee Type  Units Sold  price  revenue
0      Monday    Espresso          25   3.99    99.75
1      Monday       Latte          30   5.99   179.70
2     Tuesday    Espresso          30   3.99   119.70
3     Tuesday       Latte          20   5.99   119.80
4   Wednesday    Espresso          35   3.99   139.65
5   Wednesday       Latte          25   5.99   149.75
6    Thursday    Espresso          40   3.99   159.60
7    Thursday       Latte          30   5.99   179.70
8      Friday    Espresso          45   3.99   179.55
9      Friday       Latte          35   5.99   209.65
10   Saturday    Espresso          45   3.99   179.55
11   Saturday       Latte          35   5.99   209.65
12     Sunday    Espresso          45   3.99   179.55
13     Sunday       Latte          35   5.99   209.65

In [194]:
# lambda
coffee["Units Solds"] = coffee["price"].apply(lambda x : "Expensive" if x == 5.99 else "less expensive")


In [195]:
coffee

Day Coffee Type  Units Sold  price  revenue     Units Solds
0      Monday    Espresso          25   3.99    99.75  less expensive
1      Monday       Latte          30   5.99   179.70       Expensive
2     Tuesday    Espresso          30   3.99   119.70  less expensive
3     Tuesday       Latte          20   5.99   119.80       Expensive
4   Wednesday    Espresso          35   3.99   139.65  less expensive
5   Wednesday       Latte          25   5.99   149.75       Expensive
6    Thursday    Espresso          40   3.99   159.60  less expensive
7    Thursday       Latte          30   5.99   179.70       Expensive
8      Friday    Espresso          45   3.99   179.55  less expensive
9      Friday       Latte          35   5.99   209.65       Expensive
10   Saturday    Espresso          45   3.99   179.55  less expensive
11   Saturday       Latte          35   5.99   209.65       Expensive
12     Sunday    Espresso          45   3.99   179.55  less expensive
13     Sunday       Latte          35   5.99   209.65       Expensive

here we have added another column by using lambda where if the price is > 5.99,then its expensive else it is less expensive. Also used if we want something specific


In [213]:
# handling null value
coffee.loc[[0,1],"Units Sold"]=np.nan
coffee



Day Coffee Type  Units Sold  price  revenue     Units Solds
0      Monday    Espresso         NaN   3.99    99.75  less expensive
1      Monday       Latte         NaN   5.99   179.70       Expensive
2     Tuesday    Espresso        30.0   3.99   119.70  less expensive
3     Tuesday       Latte        20.0   5.99   119.80       Expensive
4   Wednesday    Espresso        35.0   3.99   139.65  less expensive
5   Wednesday       Latte        25.0   5.99   149.75       Expensive
6    Thursday    Espresso        40.0   3.99   159.60  less expensive
7    Thursday       Latte        30.0   5.99   179.70       Expensive
8      Friday    Espresso        45.0   3.99   179.55  less expensive
9      Friday       Latte        35.0   5.99   209.65       Expensive
10   Saturday    Espresso        45.0   3.99   179.55  less expensive
11   Saturday       Latte        35.0   5.99   209.65       Expensive
12     Sunday    Espresso        45.0   3.99   179.55  less expensive
13     Sunday       Latte        35.0   5.99   209.65       Expensive

In [214]:
coffee.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Day          14 non-null     object 
 1   Coffee Type  14 non-null     object 
 2   Units Sold   12 non-null     float64
 3   price        14 non-null     float64
 4   revenue      14 non-null     float64
 5   Units Solds  14 non-null     object 
dtypes: float64(3), object(3)
memory usage: 804.0+ bytes


In [215]:
coffee.isna().sum() #gets the null value

Day            0
Coffee Type    0
Units Sold     2
price          0
revenue        0
Units Solds    0
dtype: int64

In [216]:
# coffee = coffee.fillna(coffee["Units Sold"].mean())

In [217]:
coffee

Day Coffee Type  Units Sold  price  revenue     Units Solds
0      Monday    Espresso         NaN   3.99    99.75  less expensive
1      Monday       Latte         NaN   5.99   179.70       Expensive
2     Tuesday    Espresso        30.0   3.99   119.70  less expensive
3     Tuesday       Latte        20.0   5.99   119.80       Expensive
4   Wednesday    Espresso        35.0   3.99   139.65  less expensive
5   Wednesday       Latte        25.0   5.99   149.75       Expensive
6    Thursday    Espresso        40.0   3.99   159.60  less expensive
7    Thursday       Latte        30.0   5.99   179.70       Expensive
8      Friday    Espresso        45.0   3.99   179.55  less expensive
9      Friday       Latte        35.0   5.99   209.65       Expensive
10   Saturday    Espresso        45.0   3.99   179.55  less expensive
11   Saturday       Latte        35.0   5.99   209.65       Expensive
12     Sunday    Espresso        45.0   3.99   179.55  less expensive
13     Sunday       Latte        35.0   5.99   209.65       Expensive

In [220]:
coffee[coffee["Units Sold"].isna()] #returns datas with null values

Day Coffee Type  Units Sold  price  revenue     Units Solds
0  Monday    Espresso         NaN   3.99    99.75  less expensive
1  Monday       Latte         NaN   5.99   179.70       Expensive

In [221]:
# groupby

coffee.groupby(["Coffee Type"])["Units Sold"].mean()

Coffee Type
Espresso    40.0
Latte       30.0
Name: Units Sold, dtype: float64

In [222]:
coffee.describe()

Units Sold      price     revenue
count    12.00000  14.000000   14.000000
mean     35.00000   4.990000  165.375000
std       7.97724   1.037749   35.413192
min      20.00000   3.990000   99.750000
25%      30.00000   3.990000  142.175000
50%      35.00000   4.990000  179.550000
75%      41.25000   5.990000  179.700000
max      45.00000   5.990000  209.650000

The coffee sales dataset records daily sales of two beverage types—Espresso and Latte—across a week. Each day includes unit sales for both drinks, showing consistent trends where Espresso outperforms Latte. As the week progresses, sales for both types increase steadily, peaking on the weekend with Friday, Saturday, and Sunday showing the highest demand. This pattern suggests a growing consumer preference for coffee as the week advances, possibly tied to work fatigue or leisure activities.

I have added few columns including price and revenue as well. I have done few basic pandas operations as well.

The dataset highlights consistent sales activity across beverages, with unit sales averaging 35 and ranging from 20 to 45, indicating moderate variability. Pricing remains stable around ₹4.99, suggesting fixed rates across product types, likely Espresso and Latte. Revenue figures correlate strongly with sales volume, with most earnings clustering between ₹142.18 and ₹179.70, showcasing a healthy profit margin. Overall, the data reflects predictable pricing and reliable sales performance over time.

